File: Plugins/Kingbotk/Branches/Version 1/WikiFunctions2/LoggingBase.vb
Namespace Logging
    ''' <summary>
    ''' This abstract class can be used to build trace listener classes, or you can build a class from scratch and implement IMyTraceListener
    ''' </summary>
    Public MustInherit Class TraceListenerBase
        Inherits System.IO.StreamWriter
        Implements IMyTraceListener

        ' Initialisation
        Public Sub New(ByVal filename As String)
            MyBase.New(filename, False, System.Text.Encoding.UTF8)
        End Sub

#Region "IMyTraceListener interface"
        Public Overrides Sub Close() Implements IMyTraceListener.Close
            MyBase.Close()
        End Sub
        Public Overrides Sub Flush() Implements IMyTraceListener.Flush
            MyBase.Flush()
        End Sub
        Public MustOverride Sub ProcessingArticle(ByVal FullArticleTitle As String, ByVal NS As Namespaces) _
           Implements IMyTraceListener.ProcessingArticle
        Public MustOverride Sub WriteBulletedLine(ByVal Line As String, ByVal Bold As Boolean, _
        ByVal VerboseOnly As Boolean, ByVal DateStamp As Boolean) _
        Implements IMyTraceListener.WriteBulletedLine
        Public Sub WriteBulletedLine(ByVal Line As String, ByVal Bold As Boolean, _
        ByVal VerboseOnly As Boolean) Implements IMyTraceListener.WriteBulletedLine
            WriteBulletedLine(Line, Bold, VerboseOnly, False)
        End Sub
        Public MustOverride Sub SkippedArticle(ByVal SkippedBy As String, ByVal Reason As String) _
           Implements IMyTraceListener.SkippedArticle
        Public MustOverride Sub SkippedArticleBadTag(ByVal SkippedBy As String, ByVal FullArticleTitle As String, _
        ByVal NS As Namespaces) Implements IMyTraceListener.SkippedArticleBadTag
        Public MustOverride Sub WriteArticleActionLine(ByVal Line As String, ByVal PluginName As String) _
           Implements IMyTraceListener.WriteArticleActionLine
        Public MustOverride Sub WriteTemplateAdded(ByVal Template As String, ByVal PluginName As String) _
           Implements IMyTraceListener.WriteTemplateAdded
        Public MustOverride Sub WriteComment(ByVal Line As String) Implements IMyTraceListener.WriteComment
        Public MustOverride Sub WriteCommentAndNewLine(ByVal Line As String) _
           Implements IMyTraceListener.WriteCommentAndNewLine
        Public Overridable Sub SkippedArticleRedlink(ByVal SkippedBy As String, ByVal FullArticleTitle As String, ByVal NS As Namespaces) _
           Implements IMyTraceListener.SkippedArticleRedlink
            SkippedArticle(SkippedBy, "Attached article doesn't exist - maybe deleted?")
        End Sub
        Public Sub WriteArticleActionLineVerbose(ByVal Line As String, ByVal PluginName As String, _
        ByVal VerboseOnly As Boolean) Implements IMyTraceListener.WriteArticleActionLine
            If VerboseOnly AndAlso Not Verbose Then Exit Sub
            WriteArticleActionLine(Line, PluginName)
        End Sub
        Public MustOverride ReadOnly Property Uploadable() As Boolean Implements IMyTraceListener.Uploadable
        Public Overrides Sub Write(ByVal value As String) Implements IMyTraceListener.Write
            MyBase.Write(value)
        End Sub
        Public Overrides Sub WriteLine(ByVal value As String) Implements IMyTraceListener.WriteLine
            MyBase.WriteLine(value)
        End Sub
#End Region

        ' Protected and public members:
        Public Shared Function GetArticleTemplate(ByVal ArticleFullTitle As String, ByVal NS As Namespaces) As String
            Dim namesp As Integer
            Static reg As New Regex("( talk)?:"), reg2 As New Regex("[^:].*:")
            Dim strnamespace As String, strtitle As String, templ As String

            Select Case NS
                Case Namespaces.Main
                    Return "#{{subst:la|" & ArticleFullTitle & "}}"
                Case Namespaces.Talk
                    Return "#{{subst:lat|" & reg2.Replace(ArticleFullTitle, "") & "}}"
                Case Else
                    namesp = DirectCast(NS, Integer)
                    strnamespace = reg.Replace(WikiFunctions.Variables.Namespaces.Item(NS), "")
                    strtitle = reg2.Replace(ArticleFullTitle, "")

                    If namesp Mod 2 = 1 Then ' talk
                        templ = "lnt"
                    Else ' not talk
                        templ = "ln"
                    End If

                    Return "#{{subst:" & templ & "|" & strnamespace & "|" & strtitle & "}}"
            End Select
        End Function
        Public Shared Function GetURL(ByVal ArticleFullTitle As String) As String
            Return Variables.URL & "index.php?title=" & _
               System.Web.HttpUtility.UrlEncode(ArticleFullTitle.Replace(" ", "_"))
        End Function
        Public MustOverride ReadOnly Property Verbose() As Boolean
    End Class

    ''' <summary>
    ''' An abstract class for building auto-uploading trace listeners
    ''' </summary>
    Public MustInherit Class TraceListenerUploadableBase
        Inherits TraceListenerBase
        Implements ITraceStatusProvider

        Protected mTraceStatus As TraceStatus, mUploadSettings As UploadableLogSettings2

        Event Upload(ByVal Sender As TraceListenerUploadableBase, ByRef Success As Boolean)

        ' Initialisation:
        Public Sub New(ByVal UploadSettings As UploadableLogSettings2, ByVal TraceStatus As TraceStatus)
            MyBase.New(TraceStatus.FileName)
            mTraceStatus = TraceStatus
            mUploadSettings = UploadSettings
        End Sub

        ' Overrides & Shadowing:
        Public Overrides ReadOnly Property Uploadable() As Boolean
            Get
                Return True
            End Get
        End Property
        Public Overrides ReadOnly Property Verbose() As Boolean
            Get
                Return mUploadSettings.LogVerbose
            End Get
        End Property
        Public Overridable Shadows Sub WriteLine(ByVal Line As String, Optional ByVal CheckCounter As Boolean = True)
            MyBase.WriteLine(Line)

            mTraceStatus.LogUpload += Line & Microsoft.VisualBasic.vbCrLf
            mTraceStatus.LinesWritten += 1

            If CheckCounter Then CheckCounterForUpload()
        End Sub
        Protected Overridable ReadOnly Property IsReadyToUpload() As Boolean
            Get
                Return mTraceStatus.LinesWrittenSinceLastUpload >= mUploadSettings.UploadMaxLines
            End Get
        End Property
        Public Overridable Sub CheckCounterForUpload()
            If IsReadyToUpload Then UploadLog()
        End Sub
        Public Shadows Sub Close(ByVal Upload As Boolean)
            If Upload Then UploadLog()
            mTraceStatus.Close()
            MyBase.Close()
        End Sub

        ' Methods:
        Public Overridable Function UploadLog(Optional ByVal NewJob As Boolean = False) As Boolean
            RaiseEvent Upload(Me, UploadLog)
            ' TODO: Logging: Get result, reset TraceStatus, write success/failure to log
            If NewJob Then
                mTraceStatus.PageNumber = 1
                mTraceStatus.StartDate = Date.Now
            Else
                mTraceStatus.PageNumber += 1
                mTraceStatus.LinesWrittenSinceLastUpload = 0
                mTraceStatus.LogUpload = ""
            End If
        End Function

        ' Properties:
        Public ReadOnly Property TraceStatus() As Uploader.TraceStatus Implements _
        Uploader.ITraceStatusProvider.TraceStatus
            Get
                Return mTraceStatus
            End Get
        End Property
        Public ReadOnly Property UploadSettings() As UploadableLogSettings2
            Get
                Return mUploadSettings
            End Get
        End Property
        Public Overridable ReadOnly Property PageName() As String
            Get
                Return String.Format("{0:ddMMyy} {1}", mTraceStatus.StartDate, mUploadSettings.UploadJobName)
            End Get
        End Property
    End Class
End Namespace

File: Plugins/Kingbotk/Branches/Version 1/WF2CS/LoggingBase.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

using WikiFunctions;
using WikiFunctions.Logging;

using Logging.Uploader;

namespace Logging
{
    /// <summary>
    /// This abstract class can be used to build trace listener classes, or you can build a class from scratch and implement IMyTraceListener
    /// </summary>
    public abstract class TraceListenerBase : System.IO.StreamWriter, WikiFunctions.Logging.IMyTraceListener
    {

		// Initialisation
//INSTANT C# NOTE: These were formerly VB static local variables:
private static Regex GetArticleTemplate_reg = new Regex("( talk)?:");

		public TraceListenerBase(string filename) : base(filename, false, System.Text.Encoding.UTF8)
		{
		}

#region IMyTraceListener interface
		public override void Close()
		{
			base.Close();
		}
		public override void Flush()
		{
			base.Flush();
		}
		public abstract void ProcessingArticle(string FullArticleTitle, Namespaces NS);
		public abstract void WriteBulletedLine(string Line, bool Bold, bool VerboseOnly, bool DateStamp);
		public void WriteBulletedLine(string Line, bool Bold, bool VerboseOnly)
		{
			WriteBulletedLine(Line, Bold, VerboseOnly, false);
		}
		public abstract void SkippedArticle(string SkippedBy, string Reason);
		public abstract void SkippedArticleBadTag(string SkippedBy, string FullArticleTitle, Namespaces NS);
		public abstract void WriteArticleActionLine(string Line, string PluginName);
		public abstract void WriteTemplateAdded(string Template, string PluginName);
		public abstract void WriteComment(string Line);
		public abstract void WriteCommentAndNewLine(string Line);
		public virtual void SkippedArticleRedlink(string SkippedBy, string FullArticleTitle, Namespaces NS)
		{
			SkippedArticle(SkippedBy, "Attached article doesn't exist - maybe deleted?");
		}
		public void WriteArticleActionLine(string Line, string PluginName, bool VerboseOnly)
		{
			WriteArticleActionLineVerbose(Line, PluginName, VerboseOnly);
		}
		public void WriteArticleActionLineVerbose(string Line, string PluginName, bool VerboseOnly)
		{
			if (VerboseOnly && ! Verbose)
			{
				return;
			}
			WriteArticleActionLine(Line, PluginName);
		}
		public abstract bool Uploadable {get;}
		public override void Write(string value)
		{
			base.Write(value);
		}
		public override void WriteLine(string value)
		{
			base.WriteLine(value);
		}
#endregion

		// Protected and public members:
		public static string GetArticleTemplate(string ArticleFullTitle, Namespaces NS)
		{
			int namesp = 0;
			string strnamespace = null;
			string templ = null;
//INSTANT C# NOTE: VB local static variable moved to class level
//			Static reg As new Regex("( talk)?:")

//INSTANT C# NOTE: The following VB 'Select Case' included range-type or non-constant 'Case' expressions and was converted to C# 'if-else' logic:
//			Select Case NS
//ORIGINAL LINE: Case Namespaces.Main
			if (NS == Namespaces.Main)
			{
					return "#{{subst:la|" + ArticleFullTitle + "}}";
			}
//ORIGINAL LINE: Case Namespaces.Talk
			else if (NS == Namespaces.Talk)
			{
					return "#{{subst:lat|" + WikiFunctions.Tools.RemoveNamespaceString(ArticleFullTitle).Trim() + "}}";
			}
//ORIGINAL LINE: Case Else
			else
			{
					namesp = (int)NS;
                    strnamespace = GetArticleTemplate_reg.Replace(WikiFunctions.Variables.Namespaces.Item(NS), "");

					if (namesp % 2 == 1) // talk
					{
						templ = "lnt";
					}
					else // not talk
					{
						templ = "ln";
					}

					return "#{{subst:" + templ + "|" + strnamespace + "|" + WikiFunctions.Tools.RemoveNamespaceString(ArticleFullTitle).Trim() + "}}";
			}
		}
		public abstract bool Verbose {get;}
	}

	/// <summary>
	/// An abstract class for building auto-uploading trace listeners
	/// </summary>
    public abstract class TraceListenerUploadableBase : TraceListenerBase, ITraceStatusProvider
	{

        protected TraceStatus mTraceStatus;
        protected UploadableLogSettings2 mUploadSettings;

		public delegate void UploadEventHandler(TraceListenerUploadableBase Sender, ref bool Success);
		public event UploadEventHandler Upload;

		// Initialisation:
        public TraceListenerUploadableBase(UploadableLogSettings2 UploadSettings, TraceStatus TraceStatus)
            : base(TraceStatus.FileName)
		{
			mTraceStatus = TraceStatus;
			mUploadSettings = UploadSettings;
		}

		// Overrides & Shadowing:
		public override bool Uploadable
		{
			get
			{
				return true;
			}
		}
		public override bool Verbose
		{
			get
			{
				return mUploadSettings.LogVerbose;
			}
		}

		public virtual new void WriteLine(string Line)
		{
			WriteLine(Line, true);
		}

//INSTANT C# NOTE: C# does not support optional parameters. Overloaded method(s) are created above.
//ORIGINAL LINE: Public Overridable Shadows Sub WriteLine(ByVal Line As String, Optional ByVal CheckCounter As Boolean = true)
		public virtual new void WriteLine(string Line, bool CheckCounter)
		{
			base.WriteLine(Line);

			mTraceStatus.LogUpload += Line + System.Environment.NewLine;
			mTraceStatus.LinesWritten += 1;

			if (CheckCounter)
			{
				CheckCounterForUpload();
			}
		}
		protected virtual bool IsReadyToUpload
		{
			get
			{
				return mTraceStatus.LinesWrittenSinceLastUpload >= mUploadSettings.UploadMaxLines;
			}
		}
		public virtual void CheckCounterForUpload()
		{
			if (IsReadyToUpload)
			{
				UploadLog();
			}
		}
		public new void Close(bool Upload)
		{
			if (Upload)
			{
				UploadLog();
			}
			mTraceStatus.Close();
			base.Close();
		}

		// Methods:

		public virtual bool UploadLog()
		{
			return UploadLog(false);
		}

//INSTANT C# NOTE: C# does not support optional parameters. Overloaded method(s) are created above.
//ORIGINAL LINE: Public Overridable Function UploadLog(Optional ByVal NewJob As Boolean = false) As Boolean
		public virtual bool UploadLog(bool NewJob)
		{
			if (Upload != null)
				Upload(this, UploadLog);
			// TODO: Logging: Get result, reset TraceStatus, write success/failure to log
			if (NewJob)
			{
				mTraceStatus.PageNumber = 1;
				mTraceStatus.StartDate = System.DateTime.Now;
			}
			else
			{
				mTraceStatus.PageNumber += 1;
				mTraceStatus.LinesWrittenSinceLastUpload = 0;
				mTraceStatus.LogUpload = "";
			}
			//INSTANT C# NOTE: Inserted the following 'return' since all code paths must return a value in C#:
			return false;
		}

		// Properties:
		public Uploader.TraceStatus TraceStatus
		{
			get
			{
				return mTraceStatus;
			}
		}
		public UploadableLogSettings2 UploadSettings
		{
			get
			{
				return mUploadSettings;
			}
		}
		public virtual string PageName
		{
			get
			{
				return string.Format("{0:ddMMyy} {1}", mTraceStatus.StartDate, mUploadSettings.UploadJobName);
			}
		}
	}
}
